File: src/main/java/com/xin/server/packet/server/S_ObjectList.java
package com.xin.server.packet.server;

import java.util.List;

import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.xin.server.model.instance.NpcInstance;
import com.xin.server.packet.ServerBasePacket;
import com.xin.server.packet.ServerOpcodes;
import com.xin.server.types.NpcType;
import com.xin.server.world.WorldNpc;

/**
 * 地圖物件清單（對應 opcode {@code S_OBJECT_LIST}）。
 * <p>
 * 進入遊戲、換圖完成或前端主動請求（{@code C_OBJECT_LIST}）時推送，
 * 內含該地圖上所有 NPC／怪物／採集點／場景物件。
 * <p>
 * {@code type} 見 {@link NpcType}：0=怪物(可攻擊)、1=NPC(可交談)、2=商店(可交談買賣)、
 * 3=採集物(植物／礦物)、4=場景物件(不可互動)。
 * {@code attackable}/{@code talkable}/{@code shop}/{@code gatherable} 是由 type 推導出的
 * 便利旗標，讓前端不必重複判斷分類規則。
 *
 * <pre>
 * {
 *   "op": "S_OBJECT_LIST",
 *   "data": {
 *     "mapId": 1,
 *     "objects": [
 *       { "objId": 100001, "npcId": 45000, "name": "野狼", "type": 0, "gfxid": 1,
 *         "x": 10, "y": 12, "heading": 2,
 *         "maxHp": 50, "currentHp": 50,
 *         "attackable": true, "talkable": false, "shop": false, "gatherable": false }
 *     ]
 *   }
 * }
 * </pre>
 */
public class S_ObjectList extends ServerBasePacket {

    private S_ObjectList(int mapId) {
        super(ServerOpcodes.S_OBJECT_LIST);
        put("mapId", mapId);

        List<NpcInstance> npcs = WorldNpc.get().getNpcsByMap(mapId);
        ArrayNode arr = newArray();
        for (NpcInstance npc : npcs) {
            int type = npc.getType();
            ObjectNode node = newObject();
            node.put("objId",      npc.getId());
            node.put("npcId",      npc.getNpcTemplateId());
            node.put("name",       npc.getName());
            node.put("type",       type);
            node.put("gfxid",      npc.getGfxid());
            node.put("x",          npc.getX());
            node.put("y",          npc.getY());
            node.put("heading",    npc.getHeading());
            node.put("maxHp",      npc.getMaxHp());
            node.put("currentHp",  npc.getCurrentHp());
            node.put("attackable", NpcType.isAttackable(type));
            node.put("talkable",   NpcType.isTalkable(type));
            node.put("shop",       NpcType.isShop(type));
            node.put("gatherable", NpcType.isGatherable(type));
            arr.add(node);
        }
        putArray("objects", arr);
    }

    /**
     * 建立指定地圖的 {@link S_ObjectList} 封包。
     *
     * @param mapId 地圖編號
     */
    public static S_ObjectList of(int mapId) {
        return new S_ObjectList(mapId);
    }
}

File: lib/network/packets/server/s_object_list.dart
/// 地圖物件分類（對應伺服器 NpcType，DB npc.type_name）。
enum ObjectType {
  /// 可攻擊的怪物
  monster,

  /// 可交談的 NPC
  npc,

  /// 可交談並買賣的商店 NPC
  shop,

  /// 可互動採集的資源（植物／礦物）
  gather,

  /// 純裝飾場景物件，不可互動
  scenery;

  /// 由封包的 int 還原；未知值一律當作 [scenery]（最保守，不開放互動）。
  static ObjectType fromCode(int code) =>
      (code >= 0 && code < ObjectType.values.length)
          ? ObjectType.values[code]
          : ObjectType.scenery;
}

/// 地圖上的一個物件（NPC／怪物／採集點／場景物件）。
class MapObject {
  const MapObject({
    required this.objId,
    required this.npcId,
    required this.name,
    required this.type,
    required this.gfxid,
    required this.x,
    required this.y,
    required this.heading,
    required this.maxHp,
    required this.currentHp,
    required this.attackable,
    required this.talkable,
    required this.shop,
    required this.gatherable,
  });

  /// 世界唯一物件序號（互動封包以此指定目標）。
  final int objId;

  /// npc 模板編號。
  final int npcId;
  final String name;
  final ObjectType type;

  /// 外型編號。
  final int gfxid;
  final int x;
  final int y;

  /// 面向 0..7（0=NE,1=E,2=SE,3=S,4=SW,5=W,6=NW,7=N）。
  final int heading;
  final int maxHp;
  final int currentHp;

  /// 以下四項由伺服器依 [type] 推導，前端不需重複判斷分類規則。
  final bool attackable;
  final bool talkable;
  final bool shop;
  final bool gatherable;

  /// 是否可做任何互動（可點擊）。
  bool get interactive => type != ObjectType.scenery;

  factory MapObject.fromJson(Map<String, dynamic> j) => MapObject(
        objId: (j['objId'] as num?)?.toInt() ?? 0,
        npcId: (j['npcId'] as num?)?.toInt() ?? 0,
        name: j['name'] as String? ?? '',
        type: ObjectType.fromCode((j['type'] as num?)?.toInt() ?? 4),
        gfxid: (j['gfxid'] as num?)?.toInt() ?? 0,
        x: (j['x'] as num?)?.toInt() ?? 0,
        y: (j['y'] as num?)?.toInt() ?? 0,
        heading: (j['heading'] as num?)?.toInt() ?? 2,
        maxHp: (j['maxHp'] as num?)?.toInt() ?? 0,
        currentHp: (j['currentHp'] as num?)?.toInt() ?? 0,
        attackable: j['attackable'] as bool? ?? false,
        talkable: j['talkable'] as bool? ?? false,
        shop: j['shop'] as bool? ?? false,
        gatherable: j['gatherable'] as bool? ?? false,
      );
}

/// 地圖物件清單（回應 C_OBJECT_LIST，或進圖／換圖後由伺服器主動推送）。
class SObjectList {
  const SObjectList({required this.mapId, required this.objects});

  final int mapId;
  final List<MapObject> objects;

  factory SObjectList.fromData(Map<String, dynamic> data) => SObjectList(
        mapId: (data['mapId'] as num?)?.toInt() ?? 0,
        objects: (data['objects'] as List<dynamic>? ?? const [])
            .map((o) => MapObject.fromJson(o as Map<String, dynamic>))
            .toList(),
      );
}
